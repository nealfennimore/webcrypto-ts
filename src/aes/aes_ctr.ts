/**
 * Code related to AES_CTR mode
 * @module
 */

import * as params from "../params.js";
import * as proxy from "../proxy.js";
import { getValues } from "../random.js";
import {
    AesCtrCryptoKey,
    AesCtrProxiedCryptoKey,
    AesShared,
    Alg,
} from "./shared.js";

/** @hidden */
export const handler: ProxyHandler<AesCtrCryptoKey> = {
    get(target: AesCtrCryptoKey, prop: string) {
        switch (prop) {
            case "self":
                return target;

            case "encrypt":
                return (
                    algorithm: Omit<params.EnforcedAesCtrParams, "name">,
                    data: BufferSource
                ) => encrypt(algorithm, target, data);

            case "decrypt":
                return (
                    algorithm: Omit<params.EnforcedAesCtrParams, "name">,
                    data: BufferSource
                ) => decrypt(algorithm, target, data);

            case "wrapKey":
                return (
                    format: KeyFormat,
                    key: CryptoKey,
                    wrapAlgorithm: Omit<params.EnforcedAesCtrParams, "name">
                ) => wrapKey(format, key, target, wrapAlgorithm);

            case "unwrapKey":
                return (
                    format: KeyFormat,
                    wrappedKey: BufferSource,
                    wrappedKeyAlgorithm: params.EnforcedImportParams,
                    unwrappingKeyAlgorithm: Omit<
                        params.EnforcedAesCtrParams,
                        "name"
                    >,
                    extractable?: boolean,
                    keyUsages?: KeyUsage[]
                ) =>
                    unwrapKey(
                        format,
                        wrappedKey,
                        wrappedKeyAlgorithm,
                        target,
                        unwrappingKeyAlgorithm,
                        extractable,
                        keyUsages
                    );

            case "exportKey":
                return (format: KeyFormat) => exportKey(format, target);
        }

        return Reflect.get(target, prop);
    },
};

/**
 * Generates a counter, with the given length, starting from the count of 1. The nonce is randomized.
 * @see https://developer.mozilla.org/en-US/docs/Web/API/AesCtrParams
 * @example
 * ```ts
 * const counter = await AES_CTR.generateCounter();
 * ```
 */
export async function generateCounter(
    counterLength: number = 8
): Promise<Uint8Array> {
    const nonce = await getValues(16 - counterLength);
    const counter = new Uint8Array(16);
    counter.set([1], 15);
    counter.set(nonce);
    return counter;
}

/**
 * Generate a new AES_CTR key
 * @example
 * ```ts
 * const key = await AES_CTR.generateKey();
 * ```
 */
export async function generateKey(
    algorithm: Omit<params.EnforcedAesKeyGenParams, "name"> = {
        length: 256,
    },
    extractable: boolean = true,
    keyUsages?: KeyUsage[]
): Promise<AesCtrProxiedCryptoKey> {
    const key = await AesShared.generateKey<AesCtrCryptoKey>(
        {
            ...algorithm,
            name: Alg.Mode.AES_CTR,
        },
        extractable,
        keyUsages
    );
    return proxy.proxifyKey<AesCtrCryptoKey, AesCtrProxiedCryptoKey>(handler)(
        key
    );
}

/**
 * Import an AES_CTR key from the specified format
 * @example
 * ```ts
 * const key = await AES_CTR.importKey("jwk", jwk, { length: 256 });
 * ```
 */
export async function importKey(
    format: KeyFormat,
    key: BufferSource | JsonWebKey,
    algorithm: Omit<params.AesCtrKeyAlgorithm, "name">,
    extractable?: boolean,
    keyUsages?: KeyUsage[]
): Promise<AesCtrProxiedCryptoKey> {
    const importedKey = (await AesShared.importKey(
        format as any,
        key as any,
        {
            ...algorithm,
            name: Alg.Mode.AES_CTR,
        },
        extractable,
        keyUsages
    )) as AesCtrCryptoKey;
    return proxy.proxifyKey<AesCtrCryptoKey, AesCtrProxiedCryptoKey>(handler)(
        importedKey
    );
}

/**
 * Export an AES_CTR key into the specified format
 * @example
 * ```ts
 * const key = await AES_CTR.generateKey();
 * const jwk = await AES_CTR.exportKey("jwk", key.self);
 * ```
 * @example
 * ```ts
 * const key = await AES_CTR.generateKey();
 * const jwk = await key.exportKey("jwk");
 * ```
 */
export const exportKey = async (format: KeyFormat, key: AesCtrCryptoKey) =>
    AesShared.exportKey(format, key);

/**
 * Encrypt with an AES_CTR key
 * @example
 * ```ts
 * const key = await AES_CTR.generateKey();
 * const message = new TextEncoder().encode("a message");
 * const length = 8;
 * const counter = await AES_CTR.generateCounter(length);
 * const data = await AES_CTR.encrypt({length, counter}, key.self, message);
 * ```
 * @example
 * ```ts
 * const key = await AES_CTR.generateKey();
 * const message = new TextEncoder().encode("a message");
 * const length = 8;
 * const counter = await AES_CTR.generateCounter(length);
 * const data = await key.encrypt({length, counter}, message);
 * ```
 */
export async function encrypt(
    algorithm: Omit<params.EnforcedAesCtrParams, "name">,
    key: AesCtrCryptoKey,
    data: BufferSource
): Promise<ArrayBuffer> {
    return await AesShared.encrypt(
        {
            ...algorithm,
            name: Alg.Mode.AES_CTR,
        },
        key,
        data
    );
}

/**
 * Decrypt with an AES_CTR key
 * @example
 * ```ts
 * const data = await AES_CTR.decrypt({length, counter}, key.self, data);
 * ```
 * @example
 * ```ts
 * const data = await key.decrypt({length, counter}, data);
 * ```
 */
export async function decrypt(
    algorithm: Omit<params.EnforcedAesCtrParams, "name">,
    key: AesCtrCryptoKey,
    data: BufferSource
): Promise<ArrayBuffer> {
    return await AesShared.decrypt(
        {
            ...algorithm,
            name: Alg.Mode.AES_CTR,
        },
        key,
        data
    );
}

/**
 * Wrap another key with an AES_CTR key
 * @example
 * ```ts
 * const kek = await AES_CTR.generateKey({length: 256}, true, ['wrapKey', 'unwrapKey']);
 * const dek = await AES_CTR.generateKey();
 * const length = 8;
 * const counter = await AES_CTR.generateCounter(length);
 * const wrappedKey = await AES_CTR.wrapKey("raw", dek.self, kek.self, {length, counter});
 * ```
 * @example
 * ```ts
 * const kek = await AES_CTR.generateKey({length: 256}, true, ['wrapKey', 'unwrapKey']);
 * const dek = await AES_CTR.generateKey();
 * const length = 8;
 * const counter = await AES_CTR.generateCounter(length);
 * const wrappedKey = await kek.wrapKey("raw", dek.self, {length, counter});
 * ```
 */
export async function wrapKey(
    format: KeyFormat,
    key: CryptoKey,
    wrappingkey: AesCtrCryptoKey,
    wrapAlgorithm: Omit<params.EnforcedAesCtrParams, "name">
): Promise<ArrayBuffer> {
    return await AesShared.wrapKey(format as any, key, wrappingkey, {
        ...wrapAlgorithm,
        name: Alg.Mode.AES_CTR,
    });
}

/**
 * Unwrap a wrapped key using the key encryption key
 * @example
 * ```ts
 * const dek = await AES_CTR.unwrapKey("raw", wrappedKey, {name: "AES_CTR"}, kek.self, {length, counter});
 * ```
 * @example
 * ```ts
 * const dek = await kek.unwrapKey("raw", wrappedKey, {name: "AES_CTR"}, {length, counter});
 * ```
 */
export async function unwrapKey(
    format: KeyFormat,
    wrappedKey: BufferSource,
    wrappedKeyAlgorithm: params.EnforcedImportParams,
    unwrappingKey: AesCtrCryptoKey,
    unwrappingKeyAlgorithm: Omit<params.EnforcedAesCtrParams, "name">,
    extractable: boolean = true,
    keyUsages?: KeyUsage[]
): Promise<CryptoKey> {
    return await AesShared.unwrapKey(
        format,
        wrappedKey,
        wrappedKeyAlgorithm,
        unwrappingKey,
        {
            ...unwrappingKeyAlgorithm,
            name: Alg.Mode.AES_CTR,
        },
        extractable,
        keyUsages
    );
}
