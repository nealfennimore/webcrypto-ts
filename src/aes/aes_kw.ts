/**
 * Code related to AES_KW mode
 * @module
 */
import * as params from "../params.js";
import * as proxy from "../proxy.js";
import { AesKwCryptoKey, AesShared, Alg } from "./shared.js";

export interface AesKwProxiedCryptoKey
    extends proxy.ProxiedCryptoKey<AesKwCryptoKey> {
    wrapKey(format: KeyFormat, key: CryptoKey): Promise<ArrayBuffer>;

    unwrapKey(
        format: KeyFormat,
        wrappedKey: BufferSource,
        wrappedKeyAlgorithm: params.EnforcedImportParams,
        extractable?: boolean,
        keyUsages?: KeyUsage[]
    ): Promise<CryptoKey>;

    exportKey: (format: KeyFormat) => Promise<JsonWebKey | ArrayBuffer>;
}

const handler: ProxyHandler<AesKwCryptoKey> = {
    get(target: AesKwCryptoKey, prop: string) {
        switch (prop) {
            case "self":
                return target;

            case "wrapKey":
                return (format: KeyFormat, key: CryptoKey) =>
                    wrapKey(format, key, target);
            case "unwrapKey":
                return (
                    format: KeyFormat,
                    wrappedKey: BufferSource,
                    wrappedKeyAlgorithm: params.EnforcedImportParams,
                    extractable?: boolean,
                    keyUsages?: KeyUsage[]
                ) =>
                    unwrapKey(
                        format,
                        wrappedKey,
                        wrappedKeyAlgorithm,
                        target,
                        extractable,
                        keyUsages
                    );
            case "exportKey":
                return (format: KeyFormat) => exportKey(format, target);
        }

        return Reflect.get(target, prop);
    },
};

/**
 * Generate a new AES_KW key
 * @example
 * ```ts
 * const key = await AES_KW.generateKey();
 * ```
 */
export async function generateKey(
    algorithm: Omit<params.EnforcedAesKeyGenParams, "name"> = {
        length: 256,
    },
    extractable: boolean = true,
    keyUsages?: KeyUsage[]
): Promise<AesKwProxiedCryptoKey> {
    const key = (await AesShared.generateKey(
        {
            ...algorithm,
            name: Alg.Mode.AES_KW,
        },
        extractable,
        keyUsages
    )) as AesKwCryptoKey;
    return proxy.proxifyKey<AesKwCryptoKey, AesKwProxiedCryptoKey>(handler)(
        key
    );
}

/**
 * Import an AES_KW key from the specified format
 * @example
 * ```ts
 * const key = await AES_KW.importKey("jwk", jwk);
 * ```
 */
export async function importKey(
    format: KeyFormat,
    key: BufferSource | JsonWebKey,
    extractable?: boolean,
    keyUsages?: KeyUsage[]
): Promise<AesKwProxiedCryptoKey> {
    const importedKey = (await AesShared.importKey(
        format as any,
        key as any,
        {
            name: Alg.Mode.AES_KW,
        },
        extractable,
        keyUsages
    )) as AesKwCryptoKey;
    return proxy.proxifyKey<AesKwCryptoKey, AesKwProxiedCryptoKey>(handler)(
        importedKey
    );
}

/**
 * Export an AES_KW key into the specified format
 * @example
 * ```ts
 * const jwk = await AES_KW.exportKey("jwk", key.self);
 * ```
 * @example
 * ```ts
 * const jwk = await key.exportKey("jwk");
 * ```
 */
export const exportKey = async (format: KeyFormat, key: AesKwCryptoKey) =>
    AesShared.exportKey(format, key);

/**
 * Wrap another key with an AES_KW key
 * @example
 * ```ts
 * const kek = await AES_KW.generateKey({length: 256}, true, ['wrapKey', 'unwrapKey']);
 * const dek = await AES_GCM.generateKey();
 * const wrappedKey = await AES_KW.wrapKey("raw", dek.self, kek.self);
 * ```
 * @example
 * ```ts
 * const kek = await AES_KW.generateKey({length: 256}, true, ['wrapKey', 'unwrapKey']);
 * const dek = await AES_GCM.generateKey();
 * const wrappedKey = await kek.wrapKey("raw", dek.self);
 * ```
 */
export async function wrapKey(
    format: KeyFormat,
    key: CryptoKey,
    wrappingkey: AesKwCryptoKey
): Promise<ArrayBuffer> {
    return await AesShared.wrapKey(format as any, key, wrappingkey, {
        name: Alg.Mode.AES_KW,
    });
}

/**
 * Unwrap a wrapped key using the key encryption key
 * @example
 * ```ts
 * const dek = await AES_KW.unwrapKey("raw", wrappedKey, {name: "AES_GCM"}, kek.self);
 * ```
 * @example
 * ```ts
 * const dek = await kek.unwrapKey("raw", wrappedKey, {name: "AES_GCM"});
 * ```
 */
export async function unwrapKey(
    format: KeyFormat,
    wrappedKey: BufferSource,
    wrappedKeyAlgorithm: params.EnforcedImportParams,
    unwrappingKey: AesKwCryptoKey,
    extractable: boolean = true,
    keyUsages?: KeyUsage[]
): Promise<CryptoKey> {
    return await AesShared.unwrapKey(
        format,
        wrappedKey,
        wrappedKeyAlgorithm,
        unwrappingKey,
        {
            name: Alg.Mode.AES_KW,
        },
        extractable,
        keyUsages
    );
}
