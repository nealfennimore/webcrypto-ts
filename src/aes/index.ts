/**
 * All AES related modes and functions
 * @module
 */
export * as AES_CBC from "./aes_cbc.js";
export * as AES_CTR from "./aes_ctr.js";
export * as AES_GCM from "./aes_gcm.js";
export * as AES_KW from "./aes_kw.js";
export { Alg } from "./shared.js";
export type {
    AesCbcCryptoKey,
    AesCbcProxiedCryptoKey,
    AesCryptoKeys,
    AesCtrCryptoKey,
    AesCtrProxiedCryptoKey,
    AesGcmCryptoKey,
    AesGcmProxiedCryptoKey,
    AesKwCryptoKey,
    AesKwProxiedCryptoKey,
} from "./shared.js";
