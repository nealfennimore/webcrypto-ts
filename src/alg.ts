export namespace AES {
    export enum Mode {
        AES_CBC = "AES-CBC",
        AES_CTR = "AES-CTR",
        AES_GCM = "AES-GCM",
        AES_KW = "AES-KW",
    }

    export type Modes = `${Mode}`;
}

export namespace SHA {
    export enum Variant {
        SHA_1 = "SHA-1",
        SHA_256 = "SHA-256",
        SHA_384 = "SHA-384",
        SHA_512 = "SHA-512",
    }
    export type Variants = `${Variant}`;
    export type SecureVariants = `${Exclude<Variant, Variant.SHA_1>}`;
}

export namespace RSA {
    export enum Variant {
        RSA_OAEP = "RSA-OAEP",
        RSA_PSS = "RSA-PSS",
        RSASSA_PKCS1_v1_5 = "RSASSA-PKCS1-v1_5",
    }
    export type Variants = `${Variant}`;
}

export namespace EC {
    export enum Variant {
        ECDSA = "ECDSA",
        ECDH = "ECDH",
    }
    export type Variants = `${Variant}`;

    export enum Curve {
        P_256 = "P-256",
        P_384 = "P-384",
        P_521 = "P-521",
    }

    export type Curves = `${Curve}`;
}

export namespace Authentication {
    export enum Code {
        HMAC = "HMAC",
    }
    export type Codes = `${Code}`;
}
export namespace KDF {
    export enum Variant {
        PBKDF2 = "PBKDF2",
        HKDF = "HKDF",
    }
    export type Variants = `${Variant}`;
}
