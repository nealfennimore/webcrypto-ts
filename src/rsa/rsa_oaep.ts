import { getKeyUsagePairsByAlg, KeyUsagePairs } from "../keyUsages.js";
import * as params from "../params.js";
import { Alg as SHA } from "../sha/shared.js";
import * as WebCrypto from "../webcrypto.js";
import {
    Alg,
    RsaOaepCryptoKeyPair,
    RsaOaepPrivCryptoKey,
    RsaOaepPubCryptoKey,
    RsaShared,
} from "./shared.js";

export const generateKey = async (
    algorithm: Omit<params.EnforcedRsaHashedKeyGenParams, "name"> = {
        hash: SHA.Variant.SHA_512,
        modulusLength: 4096,
        publicExponent: new Uint8Array([0x01, 0x00, 0x01]),
    },
    extractable?: boolean,
    keyUsages?: KeyUsage[]
) =>
    (await RsaShared.generateKey(
        {
            ...algorithm,
            name: Alg.Variant.RSA_OAEP,
        },
        extractable,
        keyUsages
    )) as RsaOaepCryptoKeyPair;

export const importKey = async (
    format: KeyFormat,
    keyData: BufferSource | JsonWebKey,
    algorithm: Omit<params.EnforcedRsaHashedImportParams, "name">,
    extractable?: boolean,
    keyUsages?: KeyUsage[]
): Promise<RsaOaepPrivCryptoKey | RsaOaepPubCryptoKey> =>
    await RsaShared.importKey(
        format,
        keyData,
        { ...algorithm, name: Alg.Variant.RSA_OAEP },
        extractable,
        keyUsages
    );

export const exportKey = async (
    format: KeyFormat,
    keyData: RsaOaepPrivCryptoKey | RsaOaepPubCryptoKey
) => RsaShared.exportKey(format, keyData);

export async function encrypt(
    keyData: RsaOaepPubCryptoKey,
    plaintext: BufferSource,
    label?: params.EnforcedRsaOaepParams["label"]
): Promise<ArrayBuffer> {
    const algorithm: params.EnforcedRsaOaepParams = {
        name: Alg.Variant.RSA_OAEP,
        label,
    };
    return await WebCrypto.encrypt(algorithm, keyData, plaintext);
}

export async function decrypt(
    keyData: RsaOaepPrivCryptoKey,
    ciphertext: BufferSource,
    label?: params.EnforcedRsaOaepParams["label"]
): Promise<ArrayBuffer> {
    const algorithm: params.EnforcedRsaOaepParams = {
        name: Alg.Variant.RSA_OAEP,
        label,
    };
    return await WebCrypto.decrypt(algorithm, keyData, ciphertext);
}

export async function wrapKey(
    format: KeyFormat,
    key: CryptoKey,
    wrappingkey: RsaOaepPubCryptoKey,
    wrapAlgorithm?: Omit<params.EnforcedRsaOaepParams, "name">
): Promise<ArrayBuffer> {
    const _wrapAlgorithm: params.EnforcedRsaOaepParams = {
        ...wrapAlgorithm,
        name: Alg.Variant.RSA_OAEP,
    };
    return await WebCrypto.wrapKey(
        format as any,
        key,
        wrappingkey,
        _wrapAlgorithm
    );
}
export async function unwrapKey(
    format: KeyFormat,
    wrappedKey: BufferSource,
    wrappedKeyAlgorithm: params.EnforcedImportParams,
    unwrappingKey: RsaOaepPrivCryptoKey,
    unwrappingKeyAlgorithm: Omit<params.EnforcedRsaOaepParams, "name">,
    extractable: boolean = true,
    keyUsages?: KeyUsagePairs
): Promise<CryptoKey> {
    const _unwrappingKeyAlgorithm: params.EnforcedRsaOaepParams = {
        ...unwrappingKeyAlgorithm,
        name: Alg.Variant.RSA_OAEP,
    };
    return await WebCrypto.unwrapKey(
        format as any,
        wrappedKey,
        unwrappingKey,
        _unwrappingKeyAlgorithm,
        wrappedKeyAlgorithm,
        extractable,
        keyUsages ?? getKeyUsagePairsByAlg(wrappedKeyAlgorithm.name)
    );
}
