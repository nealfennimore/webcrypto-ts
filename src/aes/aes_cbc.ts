import * as alg from "../alg.js";
import * as params from "../params.js";
import { AesCbcCryptoKey, AesShared } from "./shared.js";

export namespace AES_CBC {
    export async function generateKey(
        algorithm: Omit<params.EnforcedAesKeyGenParams, "name"> = {
            length: 256,
        },
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<AesCbcCryptoKey> {
        return await AesShared.generateKey<AesCbcCryptoKey>(
            {
                ...algorithm,
                name: alg.AES.Mode.AES_CBC,
            },
            extractable,
            keyUsages
        );
    }

    export async function importKey(
        format: KeyFormat,
        keyData: BufferSource | JsonWebKey,
        algorithm: Omit<params.AesCbcKeyAlgorithm, "name">,
        extractable?: boolean,
        keyUsages?: KeyUsage[]
    ): Promise<AesCbcCryptoKey> {
        return await AesShared.importKey(
            format as any,
            keyData as any,
            {
                ...algorithm,
                name: alg.AES.Mode.AES_CBC,
            },
            extractable,
            keyUsages
        );
    }

    export const exportKey = AesShared.exportKey;

    export async function encrypt(
        algorithm: Omit<params.EnforcedAesCbcParams, "name">,
        keyData: AesCbcCryptoKey,
        plaintext: BufferSource
    ): Promise<ArrayBuffer> {
        return await AesShared.encrypt(
            {
                ...algorithm,
                name: alg.AES.Mode.AES_CBC,
            },
            keyData,
            plaintext
        );
    }

    export async function decrypt(
        algorithm: Omit<params.EnforcedAesCbcParams, "name">,
        keyData: AesCbcCryptoKey,
        ciphertext: BufferSource
    ): Promise<ArrayBuffer> {
        return await AesShared.decrypt(
            {
                ...algorithm,
                name: alg.AES.Mode.AES_CBC,
            },
            keyData,
            ciphertext
        );
    }

    export async function wrapKey(
        format: KeyFormat,
        key: CryptoKey,
        wrappingkey: AesCbcCryptoKey,
        wrapAlgorithm: Omit<params.EnforcedAesCbcParams, "name">
    ): Promise<ArrayBuffer> {
        return await AesShared.wrapKey(format as any, key, wrappingkey, {
            ...wrapAlgorithm,
            name: alg.AES.Mode.AES_CBC,
        });
    }
    export async function unwrapKey(
        format: KeyFormat,
        wrappedKey: BufferSource,
        wrappedKeyAlgorithm: params.EnforcedImportParams,
        unwrappingKey: AesCbcCryptoKey,
        unwrappingKeyAlgorithm: Omit<params.EnforcedAesCbcParams, "name">,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<CryptoKey> {
        return await AesShared.unwrapKey(
            format,
            wrappedKey,
            wrappedKeyAlgorithm,
            unwrappingKey,
            {
                ...unwrappingKeyAlgorithm,
                name: alg.AES.Mode.AES_CBC,
            },
            extractable,
            keyUsages
        );
    }
}
