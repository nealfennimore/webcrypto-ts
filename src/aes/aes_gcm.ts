import * as alg from "../alg.js";
import * as params from "../params.js";
import { AesKey, AesShared } from "./shared.js";

export namespace AES_GCM {
    export async function generateKey(
        algorithm: Omit<params.EnforcedAesKeyGenParams, "name"> = {
            length: 256,
        },
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<AesKey> {
        return await AesShared.generateKey(
            {
                ...algorithm,
                name: alg.AES.Mode.AES_GCM,
            },
            extractable,
            keyUsages
        );
    }

    export async function importKey(
        format: KeyFormat,
        keyData: BufferSource | JsonWebKey,
        algorithm: Omit<params.AesGcmKeyAlgorithm, "name">,
        extractable?: boolean,
        keyUsages?: KeyUsage[]
    ): Promise<AesKey> {
        return await AesShared.importKey(
            format as any,
            keyData as any,
            {
                ...algorithm,
                name: alg.AES.Mode.AES_GCM,
            },
            extractable,
            keyUsages
        );
    }

    export const exportKey = AesShared.exportKey;

    export async function encrypt(
        algorithm: Omit<params.EnforcedAesGcmParams, "name">,
        keyData: AesKey,
        plaintext: BufferSource
    ): Promise<ArrayBuffer> {
        return await AesShared.encrypt(
            {
                ...algorithm,
                name: alg.AES.Mode.AES_GCM,
            },
            keyData,
            plaintext
        );
    }

    export async function decrypt(
        algorithm: Omit<params.EnforcedAesGcmParams, "name">,
        keyData: AesKey,
        ciphertext: BufferSource
    ): Promise<ArrayBuffer> {
        return await AesShared.decrypt(
            {
                ...algorithm,
                name: alg.AES.Mode.AES_GCM,
            },
            keyData,
            ciphertext
        );
    }

    export async function wrapKey(
        format: KeyFormat,
        key: CryptoKey,
        wrappingkey: AesKey,
        wrapAlgorithm: Omit<params.EnforcedAesGcmParams, "name">
    ): Promise<ArrayBuffer> {
        return await AesShared.wrapKey(format as any, key, wrappingkey, {
            ...wrapAlgorithm,
            name: alg.AES.Mode.AES_GCM,
        });
    }
    export async function unwrapKey(
        format: KeyFormat,
        wrappedKey: BufferSource,
        wrappedKeyAlgorithm: params.EnforcedImportParams,
        unwrappingKey: AesKey,
        unwrappingKeyAlgorithm: Omit<params.EnforcedAesGcmParams, "name">,
        extractable: boolean = true,
        keyUsages?: KeyUsage[]
    ): Promise<CryptoKey> {
        return await AesShared.unwrapKey(
            format,
            wrappedKey,
            wrappedKeyAlgorithm,
            unwrappingKey,
            {
                ...unwrappingKeyAlgorithm,
                name: alg.AES.Mode.AES_GCM,
            },
            extractable,
            keyUsages
        );
    }
}
