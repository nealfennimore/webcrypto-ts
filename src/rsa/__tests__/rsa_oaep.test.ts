import { AES } from "../../aes/index";
import * as alg from "../../alg";
import { RSA_OAEP } from "../rsa_oaep";

describe("RSA_OAEP", () => {
    it("should generate key", async () => {
        const key = await RSA_OAEP.generateKey();
        expect(key).toMatchSnapshot();
    });

    it("should import and export key", async () => {
        const keyPair = await RSA_OAEP.generateKey();
        const jwk = await RSA_OAEP.exportKey("jwk", keyPair.publicKey);
        const pubKey = await RSA_OAEP.importKey("jwk", "SHA-512", jwk, false, [
            "encrypt",
        ]);
        const text = encode("a message");
        const ciphertext = await RSA_OAEP.encrypt(pubKey, text);
        const plaintext = await RSA_OAEP.decrypt(
            keyPair.privateKey,
            ciphertext
        );
        expect(decode(plaintext)).toEqual(decode(text));
    });
    it("should encrypt and decrypt", async () => {
        const keyPair = await RSA_OAEP.generateKey();
        const text = encode("a message");
        const ciphertext = await RSA_OAEP.encrypt(keyPair.publicKey, text);
        const plaintext = await RSA_OAEP.decrypt(
            keyPair.privateKey,
            ciphertext
        );
        expect(decode(plaintext)).toEqual(decode(text));
    });
    it("should wrap and unwrap key", async () => {
        const aesKey = await AES.AES_CBC.generateKey();
        const keyPair = await RSA_OAEP.generateKey(
            "SHA-512",
            4096,
            new Uint8Array([0x01, 0x00, 0x01]),
            true,
            ["wrapKey", "unwrapKey"]
        );

        const wrappedAesKey = await RSA_OAEP.wrapKey(
            "raw",
            aesKey,
            keyPair.publicKey,
            {}
        );
        const unwrappedAesKey = await RSA_OAEP.unwrapKey(
            "raw",
            wrappedAesKey,
            { name: alg.AES.Mode.AES_CBC },
            keyPair.privateKey,
            {}
        );

        expect(await AES.AES_CBC.exportKey("jwk", aesKey)).toEqual(
            await AES.AES_CBC.exportKey("jwk", unwrappedAesKey)
        );
    });
});
