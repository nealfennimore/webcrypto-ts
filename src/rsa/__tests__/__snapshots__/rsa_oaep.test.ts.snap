// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RSA_OAEP Original should generate key 1`] = `
Object {
  "privateKey": CryptoKey {
    Symbol(kKeyObject): PrivateKeyObject {
      Symbol(kKeyType): "private",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSA-OAEP",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "decrypt",
    ],
  },
  "publicKey": CryptoKey {
    Symbol(kKeyObject): PublicKeyObject {
      Symbol(kKeyType): "public",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSA-OAEP",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "encrypt",
    ],
  },
}
`;

exports[`RSA_OAEP Proxied should generate key 1`] = `
Object {
  "privateKey": CryptoKey {
    Symbol(kKeyObject): PrivateKeyObject {
      Symbol(kKeyType): "private",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSA-OAEP",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "decrypt",
    ],
  },
  "publicKey": CryptoKey {
    Symbol(kKeyObject): PublicKeyObject {
      Symbol(kKeyType): "public",
    },
    Symbol(kAlgorithm): Object {
      "hash": Object {
        "name": "SHA-512",
      },
      "modulusLength": 4096,
      "name": "RSA-OAEP",
      "publicExponent": Uint8Array [
        1,
        0,
        1,
      ],
    },
    Symbol(kExtractable): true,
    Symbol(kKeyUsages): Array [
      "encrypt",
    ],
  },
}
`;
